### ISSUE TITLE
Manual optimizer step, forward pass of model does not change

### ISSUE THREAD

**Issue by mhun1** [PyTorchLightning/pytorch-lightning#7128]
Hello!

In the training step I try to manually optimize the network and obtain the prediction of the updated network afterwards, but in the same training step. However, if I obtain the new prediction it is equal to the prediction of the network before the optimization. I obtained the state_dicts before and after the optimization and noticed that it changes, but if I forward through the optimized network I receive the same loss as before. (Note: It is not because of the low change of values, I tried higher learning rates resulting in high values, which lead to the same result, that the loss stayed equally)

Does anyone has an idea on this one?

See the code here:

```
def training_step(self,x,y):
        opt = self.optimizers(use_pl_optimizer=True)
        opt.zero_grad()
        pred = self.net(x)
        loss = self.loss(pred,y)
        
        params = self.net.state_dict()
        print("------------------1-------------------------")
        print(params['encoder.encode.0.conv1.conv.weight'][0][0])
        print("Loss before: ", loss.item())
        self.manual_backward(loss, opt)
        opt.step()

        params_ = self.net.state_dict()
        pred_0 = self.net(x)
        loss_2 = self.loss(pred_0, y)
        print("-----------2-----------")
        print(params_['encoder.encode.0.conv1.conv.weight'][0][0])
        print("Loss after: ", loss_2.item())
```
Output:

```
------------------1-------------------------
tensor([[[ 0.0594, -0.0953, -0.1361],
         [-0.0064,  0.0601,  0.1044],
         [-0.1681,  0.0981, -0.1655]],

        [[ 0.1407, -0.1374, -0.0645],
         [ 0.1218, -0.1436, -0.1501],
         [ 0.1565, -0.0013,  0.0699]],

        [[-0.1309,  0.0975, -0.1478],
         [ 0.1309, -0.1202,  0.1641],
         [-0.1893,  0.1068, -0.1260]]], device='cuda:0')
Loss before:  0.8495523929595947
-----------2-----------
tensor([[[ 0.0604, -0.0943, -0.1351],
         [-0.0054,  0.0611,  0.1054],
         [-0.1671,  0.0991, -0.1645]],

        [[ 0.1417, -0.1364, -0.0655],
         [ 0.1228, -0.1426, -0.1491],
         [ 0.1575, -0.0003,  0.0709]],

        [[-0.1299,  0.0985, -0.1488],
         [ 0.1319, -0.1192,  0.1631],
         [-0.1883,  0.1078, -0.1270]]], device='cuda:0')
Loss after:  0.8495523929595947
```

**Comment by akihironitta** [comment]
> (Note: It is not because of the low change of values...

@mhun1 I cannot think of any other reason than that because the model parameters seem correctly updated. To investigate further, could you provide a reproducible script and the information of your environment which you can get by running the following commands?

```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
python collect_env_details.py
```

**Comment by mhun1** [comment]
First, thanks for answering!
For the script, I need quite sometime to reproduce, since it is not all trivial.

Here is the environment information:
```
* CUDA:
	- GPU:
		- GeForce RTX 3090
	- available:         True
	- version:           11.0
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1+cu110
	- pytorch-lightning: 1.1.5
	- tqdm:              4.56.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.8.5
	- version:           #56~20.04.1-Ubuntu SMP Mon Apr 12 21:46:35 UTC 2021
```

Further, I noticed that when the training step continues, the network predicts a other loss, but with the same state_dict, I obtained in the previous training step:

```
------------------1-------------------------
tensor([[[-0.0825, -0.0832,  0.1397],
         [ 0.1788, -0.0568, -0.1567],
         [-0.1195,  0.1364,  0.1240]],

        [[ 0.1072, -0.0583, -0.0968],
         [ 0.0116,  0.0098,  0.0753],
         [ 0.0629,  0.0499,  0.0036]],

        [[ 0.1369,  0.0436, -0.0003],
         [ 0.1527,  0.0976,  0.1016],
         [-0.1591,  0.1265, -0.0731]]], device='cuda:0')
Loss before:  0.7147183418273926
-----------2-----------
tensor([[[-0.0835, -0.0842,  0.1387],
         [ 0.1778, -0.0578, -0.1577],
         [-0.1205,  0.1354,  0.1230]],

        [[ 0.1062, -0.0593, -0.0958],
         [ 0.0106,  0.0108,  0.0763],
         [ 0.0619,  0.0509,  0.0046]],

        [[ 0.1379,  0.0446,  0.0007],
         [ 0.1537,  0.0986,  0.1026],
         [-0.1581,  0.1275, -0.0721]]], device='cuda:0')
Loss after:  0.7147183418273926
Epoch 0:   7%|█████▌                                                                             | 1/15 [00:02<00:34,  2.46s/it, v_num=g4d8]working
------------------1-------------------------
tensor([[[-0.0835, -0.0842,  0.1387],
         [ 0.1778, -0.0578, -0.1577],
         [-0.1205,  0.1354,  0.1230]],

        [[ 0.1062, -0.0593, -0.0958],
         [ 0.0106,  0.0108,  0.0763],
         [ 0.0619,  0.0509,  0.0046]],

        [[ 0.1379,  0.0446,  0.0007],
         [ 0.1537,  0.0986,  0.1026],
         [-0.1581,  0.1275, -0.0721]]], device='cuda:0')
Loss before:  0.5142204761505127
```

**Comment by akihironitta** [comment]
> Further, I noticed that when the training step continues, the network predicts a other loss, but with the same state_dict, I obtained in the previous training step:

@mhun1 I don't think I understand your point, but I think it is working as intended because the model parameters get updated only at `opt.step()` in your `training_step()` in manual optimization.

It'd be wonderful if you could share your full script with us to investigate it further.
